tests/integration/tests/conftest.py

```diff
@@ -99,6 +101,15 @@ def node_count(request) -> int:
     return int(node_count_arg)
 
 
+def snap_versions(request) -> Iterator[Optional[str]]:
+    """An endless iterable of snap versions for each node in the test."""
+    marking = ()
+    if snap_version_marker := request.node.get_closest_marker("snap_versions"):
+        marking, *_ = snap_version_marker.args
+    # endlessly repeat of the configured snap version after exhausting the marking
+    return itertools.chain(marking, itertools.repeat(None))
```

Comment:
alright -- so lets clear that up here.  The developer may mark a test with the versions they desire and the node count they desire 

```python
@pytest.mark.node_count(15)
@pytest.mark.snap_versions(["1.31/stable", "1.31/edge", "1.30/stable", ...])
def my_test(instances):
    ...
```

these two markings are not congruent, meaning you can say you want 15 nodes but only specify you'd like 3 snap versions for those instances. We know the first 3 instances should be running from specific channels but how about the other instances?

Let's look at hos `setup_k8s_snap` works:
* `setup_k8s_snap` method takes a `snap` argument.  If it's non-None, the specified version will be installed on the instance.  If it's `None`, it will install from `config.SNAP` -- an environment variable defined when running the test.

So now you see this little gem of code you've highlighted with `itertools.chain(marking, itertools.repeat(None)`

will empty the iterator provided by marking.  And when it's empty it will repeat `None` forever... 

If the test is unmarked with snap_versions, this method spits out `None` infinitely
If the test is marked with snap_versions, this method splits out all the snao versions, until they're consumed, then infinite  `None`


How does that combine with python's `zip`?  Well `zip` will keep consume the next item from all the iterators it's given as a tuple.  It's like taking 1 card off of multiple decks of card and creating a hand from them.  Once a hand is created, it will take 1 card of the decks again for another hand.  Zip will do this until the first deck is consumed and then stop.  

since `snap_versions` is an infinite deck, `zip` stops when the `range(node_count)` is reached.

Reply:
Got it. Wonderful. Thanks for the thorough explanation. I was confused why the remaining elements of the chain are `None`, and now I know that's because we WANT them to be (so that we can read the version from `config.SNAP`). 
Amazing. Thank you!